src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,186 @@
+    channel->addMessage(makeSystemMessage(QString("Started upload...")));
+
+    if (source->hasFormat("image/png"))
+    {
+        uploadImageToNuuls({source->data("image/png"), "png"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/jpeg"))
+    {
+        uploadImageToNuuls({source->data("image/jpeg"), "jpeg"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/gif"))
+    {
+        uploadImageToNuuls({source->data("image/gif"), "gif"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasUrls())
+    {
+        for (const QUrl &path : source->urls())
+        {
+            if (!getImageFileFormat(path.toLocalFile()).isEmpty())
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading image: %1").arg(path.toLocalFile())));
+                QImage img = QImage(path.toLocalFile());
+                if (img.isNull())
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Couldn't load image :(")));
+                    return;
+                }
+                QByteArray imageData;
+                QBuffer buf(&imageData);
+                buf.open(QIODevice::WriteOnly);
+                img.save(&buf, "png");
+
+                TypedBytes data = {imageData, "png"};
+                uploadQueue.push(data);
+            }
+            else if (path.toLocalFile().endsWith(".gif"))
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading GIF: %1").arg(path.toLocalFile())));
+                QFile file(path.toLocalFile());
+                bool isOkay = file.open(QIODevice::ReadOnly);
+                if (!isOkay)
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Failed to open file. :(")));
+                    return;
+                }
+                TypedBytes data = {file.readAll(), "gif"};
+                uploadQueue.push(data);
+                file.close();
+                // file.readAll() => might be a bit big but it /should/ work
+            }
+            else
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Cannot upload file: %1, not an image")
+                        .arg(path.toLocalFile())));
+                isUploading = false;
+            }
+        }
+        if (uploadQueue.size())
+        {
+            uploadImageToNuuls(uploadQueue.front(), channel, outputTextEdit);
+            uploadQueue.pop();
+        }
+    }
+    else
+    {  // not PNG, try loading it into QImage and save it to a PNG.
+        QImage image = qvariant_cast<QImage>(source->imageData());
+        QByteArray imageData;
+        QBuffer buf(&imageData);
+        buf.open(QIODevice::WriteOnly);
+        image.save(&buf, "png");
```

Comment:
duplicate "png-conversion code", make a helper function in the anonymous namespace